src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -155,21 +165,22 @@ public CompletionStage<UStatus> unsubscribe(UUri topic, UListener listener, Call
 
         final UnsubscribeRequest unsubscribeRequest = UnsubscribeRequest.newBuilder().setTopic(topic).build();
         
-        CompletionStage<RpcResult<UnsubscribeResponse>> result = RpcMapper.mapResponseToResult(
+        return RpcMapper.mapResponseToResult(
             // Send the unsubscribe request
             rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options),     
                 UnsubscribeResponse.class)
             // Then unregister the listener
-            .whenComplete((response, ex) -> mHandlers.remove(topic));
-
-        // Create a CompletionStage that will handle the response from the unsubscribe request
-        return transport.unregisterListener(topic, listener)
-            .thenCombine(result, (status, response) -> {
+            .thenCompose( response ->  {
                 if (response.isSuccess()) {
-                    return UStatus.newBuilder().setCode(UCode.OK).build();
+                    return transport.unregisterListener(topic, listener);
                 }
-                return response.failureValue();
-            });
+                return CompletableFuture.completedFuture(response.failureValue());
+            })
+            .whenComplete((status, exception) -> {
+                if (status.getCode() == UCode.OK) {
+                    mHandlers.remove(topic);
+                }
+            });       
```

Comment:
if invokemethod success but unregisterlistener failed during unsubscribe, then client won't be able to remove handler for that topic anymore, hence not able to add new handler during resubscribe.
we probably should go back to previous implementation, remove handler regardless the result of unregisterlistener

Reply:
In this, I want to add that we should remove the handler only when the unsubscription is successful, without checking the unregisterListener API response

Reply:
done thank you